examples/solvers/pdhg_denoising_L1_HuberTV.py

```diff
@@ -0,0 +1,123 @@
+"""Total variation denoising using PDHG.
+
+Solve the L1-HuberTV problem
+
+        min_{x >= 0} ||x - d||_1 + lam TV_gamma(x)
+
+where ``grad`` the spatial gradient and ``d`` is given noisy data.
+
+For further details and a description of the solution method used, see
+https://odlgroup.github.io/odl/guide/pdhg_guide.html in the ODL documentation.
+"""
+
+import numpy as np
+import scipy
+import odl
+import matplotlib.pyplot as plt
+
+# --- define setting --- #
+
+# Read test image: use only every second pixel, convert integer to float
+image = np.rot90(scipy.misc.ascent()[::2, ::2].astype('float'), 3)
+shape = image.shape
+
+# Rescale max to 1
+image /= image.max()
+
+# Discretized spaces
+space = odl.uniform_discr([0, 0], shape, shape)
+
+# Create space element of ground truth
+orig = space.element(image.copy())
+
+# Add noise and convert to space element
+noisy = odl.phantom.salt_pepper_noise(orig)
+
+# Gradient operator
+gradient = odl.Gradient(space, method='forward')
+
+# regularization parameter
+reg_param = 1
+
+# l1 data matching
+l1_norm = 1 / reg_param * odl.solvers.L1Norm(space).translated(noisy)
+
+# HuberTV-regularization
+huber_l1_norm = odl.solvers.HuberL1L2(gradient.range, gamma=.1)
+
+# define objective
+obj_fun = l1_norm + huber_l1_norm * gradient
+
+# strong convexity of "f*"
+strong_convexity = 1 / huber_l1_norm.grad_lipschitz
+
+
+# define callback to store function values
+class CallbackStore(odl.solvers.util.callback.Callback):
+    def __init__(self):
+        self.iteration_count = 0
+        self.iteration_counts = []
+        self.obj_function_values = []
+
+    def __call__(self, x):
+        self.iteration_count += 1
+        self.iteration_counts.append(self.iteration_count)
+        self.obj_function_values.append(obj_fun(x))
+
+    def reset(self):
+        self.iteration_count = 0
+        self.iteration_counts = []
+        self.obj_function_values = []
+
+
+callback = (odl.solvers.CallbackPrintIteration() & CallbackStore())
+
+# number of iterations
+niter = 500
+
+# Assign operator and functionals
+op = gradient
+f = huber_l1_norm
+g = l1_norm
+
+# The operator norm of the gradient with forward differences is well-known
```

Comment:
feel free to add this to `Gradient.norm` and instead call `op_norm = gradient.norm()` here, that would be much better!

Reply:
I agree. However, the generality of the gradient operator makes this rather difficult. For forward and backward differences on some regular space we can do it. Not sure how to implement it such that only the "known" cases are covered. Maybe you can make a start with the case differentiation and I add the formulas?